xarray/core/duck_array_ops.py

```diff
@@ -372,43 +372,62 @@ def _datetime_nanmin(array):
 
 
 def datetime_to_numeric(array, offset=None, datetime_unit=None, dtype=float):
-    """Convert an array containing datetime-like data to an array of floats.
+    """Return an array containing datetime-like data to numerical values.
+
+    Convert the datetime array to a timedelta relative to an offset.
 
     Parameters
     ----------
-    da : np.array
-        Input data
-    offset: Scalar with the same type of array or None
-        If None, subtract minimum values to reduce round off error
-    datetime_unit: None or any of {'Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms',
-        'us', 'ns', 'ps', 'fs', 'as'}
-    dtype: target dtype
+    da : array-like
+      Input data
+    offset: None, datetime or cftime.datetime
+      Datetime offset. If None, this is set by default to the array's minimum
+      value to reduce round off errors.
+    datetime_unit: {None, Y, M, W, D, h, m, s, ms, us, ns, ps, fs, as}
+      If not None, convert output to a given datetime unit. Note that some
+      conversions are not allowed due to non-linear relationships between units.
+    dtype: dtype
+      Output dtype.
 
     Returns
     -------
     array
+      Numerical representation of datetime object relative to an offset.
+
+    Notes
+    -----
+    Some datetime unit conversions won't work, for example from days to years, even
+    though some calendars would allow for them (e.g. no_leap). This is because there
+    is no `cftime.timedelta` object.
     """
     # TODO: make this function dask-compatible?
+    # Set offset to minimum if not given
     if offset is None:
         if array.dtype.kind in "Mm":
             offset = _datetime_nanmin(array)
         else:
             offset = min(array)
+
+    # Compute timedelta object.
+    # For np.datetime64, this can silently yield garbage due to overflow.
     array = array - offset
 
-    if not hasattr(array, "dtype"):  # scalar is converted to 0d-array
+    # Scalar is converted to 0d-array
+    if not hasattr(array, "dtype"):
         array = np.array(array)
 
+    # Convert timedelta objects to timedelta64[ms] dtype.
     if array.dtype.kind in "O":
-        # possibly convert object array containing datetime.timedelta
-        array = np.asarray(pd.Series(array.ravel())).reshape(array.shape)
+        array = array.astype("timedelta64[ms]")
```

Comment:
Here we may want to use microsecond units (`"timedelta64[us]"`); using millisecond resolution may lead to unnecessary truncation of precision.

Reply:
Will do. In `get_clean_interp_index`, should we also use `datetime_to_numeric` for pd.DatetimeIndex with the us resolution ?

```
    if isinstance(index, (CFTimeIndex, pd.DatetimeIndex)):
        offset = type(index[0])(1970, 1, 1)
        index = Variable(
            data=datetime_to_numeric(index, offset=offset, datetime_unit="us"),
            dims=(dim,),
        )
```

Reply:
Switching to "us" breaks `interpolate_na` because max_gap is ns by default. 

Throughout xarray, "ns" is the default resolution. I'm a bit worried that switching this for "us" in a couple of places it a recipe for future bugs.

Reply:
It occurs to me that we *could* use nanosecond units (or units even finer) for cftime dates, so long as we do the unit conversion step after converting to floats.  E.g. one could write a function like the following to aid in that:
```python
SUB_MICROSECOND_UNIT_CONVERSION_FACTORS = {
    "ns": 1e3,
    "ps": 1e6,
    "fs": 1e9,
    "as": 1e12
}

def _py_timedelta_to_float(array, datetime_unit="ns"):
    array = array.astype("timedelta64[us]")
    if datetime_unit in SUB_MICROSECOND_UNIT_CONVERSION_FACTORS:
        factor = SUB_MICROSECOND_UNIT_CONVERSION_FACTORS[datetime_unit]
        return factor * array.astype(np.float64)
    else:
        return array / np.timedelta64(1, datetime_unit)
```
I have not thought carefully yet about how to handle `max_gap` in `interpolate_na` in the case of cftime dates, but this seems like a possible compromise to allow us to continue to default to nanosecond resolution internally for these functions.  @huard do you think that could help?

Reply:
Here's perhaps a cleaner version:
```python
def _py_timedelta_to_float(array, datetime_unit="ns"):
    array = array.astype("timedelta64[us]").astype(np.float64)
    conversion_factor = np.timedelta64(1, "us") / np.timedelta64(1, datetime_unit)
    return conversion_factor * array
```

Reply:
> max_gap in interpolate_na in the case of cftime dates,

It should just work if you transform `max_gap` in the same way you transform the coordinate.

Reply:
I don't have a complete enough picture to have a strong opinion on this. My feeling is that on the long term, xarray would probably want to let users pick their own time resolution. So for example when you call `xr.cftime_range`, you would have a `dtype` or `precision` argument that would be used every time objects need to be converted to numerical values. 

One potential issue with `_py_timedelta_to_float` is that we won't be able to convert back the float to the original timedelta without first casting to int, and I would suppose a loss of precision. Also you'd have to remember the original resolution. My sense is that using datetime64 and timedelta64 as the default numerical dtypes would be less error prone. 

I have a naive fix for max_gap in PR #3631 simply switching the resolution to us. Want me to merge both PRs into one ?

Reply:
> My feeling is that on the long term, xarray would probably want to let users pick their own time resolution. 

Can you elaborate on what the example use-case of this would be?  Fundamentally, without upstream changes, the underlying date objects will always be of either nanosecond or microsecond resolution.

While not exactly a "resolution" option, we do already expose `datetime_unit` arguments in relevant functions, e.g. in [differentiate](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.differentiate.html) and [integrate](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.integrate.html).  Is that along the lines of what you meant?

> One potential issue with _py_timedelta_to_float is that we won't be able to convert back the float to the original timedelta without first casting to int, and I would suppose a loss of precision. Also you'd have to remember the original resolution. 

I believe one could say the same thing about the current implementation of `datetime_to_numeric` (it ultimately produces unitless floats too).  How would you suggest rewriting that?

Reply:
> Can you elaborate on what the example use-case of this would be? Fundamentally, without upstream changes, the underlying date objects will always be of either nanosecond (mainly a constraint of pandas) or microsecond (cftime) resolution.

Agreed, but I assume upstream libraries will eventually leverage datetime64 and timedelta64 objects and become more flexible in terms of their internal resolution. This would let people use xarray to look at femtoseconds experiments in optics for example. 

> While not exactly a "resolution" option, we do already expose `datetime_unit` arguments in relevant functions, e.g. in [differentiate](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.differentiate.html) and [integrate](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.integrate.html). Is that along the lines of what you meant?

Yes, but I feel like this resolution should be a property of the underlying object. 

> I believe one could say the same thing about the current implementation of `datetime_to_numeric` (it ultimately produces unitless floats too). How would you suggest rewriting that?

Correct. I'm wondering if using timedelta64 and datetime64 objects would be cleaner. I'm not saying this should be done now though.

Reply:
Those are all good big-picture discussion points @huard.  I'll briefly add my two cents, but I think we should probably get back to the matter at hand.  It would be great to fix these interpolation bugs.

> Agreed, but I assume upstream libraries will eventually leverage datetime64 and timedelta64 objects and become more flexible in terms of their internal resolution. This would let people use xarray to look at femtoseconds experiments in optics for example.

Yes, that would be nice.  We've been down this road to some extent before; [we've discussed](https://github.com/pydata/xarray/issues/789) (see also [this mailing list thread](https://groups.google.com/forum/#!topic/pydata/kk04maBGw1U)) potentially making changes upstream in pandas to allow us to use coarser resolution datetime64 types in DatetimeIndexes for long time series, but the situation there is complicated: https://github.com/pandas-dev/pandas/issues/7307.  I'm afraid I'm not aware of any ongoing effort to change things.

> Yes, but I feel like this resolution should be a property of the underlying object.

It sounds like in this context by "resolution" you are referring to the units (as opposed to a truncation in precision), which could be more tractable.  This perhaps could be something we could discuss within the context of making xarray data structures/operations units-aware (e.g. through `pint`).  There is a fair bit of ongoing work in this area; see #3594 for a jumping off point.

> Correct. I'm wondering if using timedelta64 and datetime64 objects would be cleaner. I'm not saying this should be done now though.

I see.  Eventually for operations like interpolation and differentiation I think we need to convert to floats.  We're for sure open to ideas for how to improve things though.

Reply:
So happy to get back on target and fix this bug. What both PRs do (and I could merge them into one) is enforce the "us" precision in places where we need a numerical representation resolution in `duck_array_ops.datetime_to_numeric`, the converting to float: 
```
if array.dtype.kind in "O":
     array = array.astype("timedelta64[us]")
...
return array.astype(dtype) 
```
where dtype is `float` by default. 

I leverage that in `missing.get_clean_interp_index` with a new if clause:
```
    if isinstance(index, (CFTimeIndex, pd.DatetimeIndex)):
        offset = type(index[0])(1970, 1, 1)
        index = Variable(
            data=datetime_to_numeric(index, offset=offset, datetime_unit="us"),
            dims=(dim,),
        )
...
index = index.values.astype(np.float64)
```
This however break `interpolate_na`, so I convert `max_gap` in `missing.interp_na` to "us" before converting to float. 
```
max_gap = np.timedelta64(max_gap, "us").astype(np.float64)
```
So this change would break people's code that assume `datetime_to_numeric` and `get_clean_interp_index` return "ns" based floats. If that's unacceptable, then we could use your suggested `_py_timedelta_to_float`.

Reply:
@huard These are private functions so that's not a concern. We do want to maintain compatibility with numpy>= 0.14 though

Reply:
Yes.  Keep in mind though the default units for `differentiate` and `integrate` have been nanoseconds, and that is user-facing; should we preserve that?  Also should we preserve the ability to specify sub-microsecond `max_gap`'s in `interpolate_na`?

Reply:
ah I see. Thanks @spencerkclark 

> Keep in mind though the default units for differentiate and integrate have been nanoseconds, and that is user-facing; should we preserve that?

Yes I think we should. Or we have to go through a deprecation cycle.

> Also should we preserve the ability to specify sub-microsecond max_gap's in interpolate_na

From a user perspective, it seems like we should allow all units that may be used for the time coordinate. I admit I don't understand the subtleties of precision though.

Reply:
Thanks @dcherian -- I agree on all points.   

So @huard, I think it makes sense to do the following (feel free to merge your PRs or not as you see fit):

1. Maintain the ability to output floats with nanosecond units (or finer) from `datetime_to_numeric` by converting timedelta-like objects to floats before doing the unit conversion.
2. Use a similar unit conversion approach when converting `max_gap` to a float, so that we again get the best of both worlds.  I.e. continue to support nanosecond-resolution `max_gap` timedeltas, but at the same time allow for > 292-year `max_gap` timedeltas (specified, e.g., using `datetime.timedelta` objects or `np.timedelta64` objects of microsecond or coarser resolution).

(2) might require some careful logic, but in principle I think it would be doable.  Does that make sense?  Thanks already for your help in identifying/sorting all these issues out.  Sorry for taking a while to think all this through.

Reply:
I've merged my two PRs into #3631 and implemented the `py_timedelta_to_float` function. 
It seems to work except for earlier numpy versions. I've special-cased earlier numpy versions since I don't think I can fix that.

Reply:
Thanks a lot @huard -- I'll have a closer look over the coming days.